src/com/mendix/widget/StarRating/components/StarRating.ts

```diff
@@ -0,0 +1,134 @@
+                }),
+                createElement(Alert, { message: this.state.errorMessage }));
+        } else {
+            return createElement(Alert, { message: this.state.errorMessage });
+        }
+    }
+
+    private getRate() {
+        // The contextObject entityType for "average" is different from that of "single"
+        // InitialRate for rateType "average" comes from averageAttribute on campaignEntity
+        // InitialRate for rateType "single" comes from rateAttribute on rateEntity
+        const initialRate = this.contextObject
+            ? this.props.rateType === "average"
+                ? this.contextObject.get(this.props.averageAttribute) as number
+                : this.contextObject.get(this.props.rateAttribute) as number
+            : 0 ;
+
+        const maximumValue = this.step * this.stop;
+        if (initialRate > maximumValue) {
+            return maximumValue;
+        } else if (initialRate < this.start) {
+            return this.start;
+        } else {
+            // This helps to round off to the nearest fraction.
+            // eg fraction 2 or 0.5, rounds off a rate 1.4 to 1.5, 1.2 to 1.0
+            return Math.round(initialRate * this.fractions) / this.fractions;
+        }
+    }
+
+    private submitData(rate: number) {
+        if (this.contextObject) {
+            this.contextObject.set(this.props.rateAttribute, rate);
+            if (this.props.onChangeMicroflow) {
+                this.executeMicroflow(this.contextObject.getGuid(), this.props.onChangeMicroflow);
+            }
+        }
+    }
+
+    private executeMicroflow(mendixGUID: string, microflow: string) {
+        window.mx.ui.action(microflow, {
+            error: (error: Error) =>
+                this.setState({ errorMessage: `Error while executing microflow: ${microflow}: ${error.message}` }),
+            params: {
+                applyto: "selection",
+                guids: [ mendixGUID ]
+            }
+        });
+    }
+
+    private hasValidConfig(props: StarRatingProps ) {
+        const errorMessage: string[] = [];
+
+        if (!this.contextObject) {
+            return true;
+        }
+        if ((this.rateType === "average") && this.contextObject.getEntity() !== this.campaignEntity) {
+            errorMessage.push(" - For rate type 'average', the contextObject should be campaign entity");
+        }
+        if ((this.rateType === "single") && this.contextObject.getEntity() !== this.props.rateEntity) {
+            errorMessage.push(` - For rate type 'Single', the contextObject be rate entity '${this.props.rateEntity}'`);
+        }
+        if (this.rateType === "single" && !this.contextObject.isReference(this.ownerReference)) {
+            errorMessage.push(` - Context object has no User / Owner association to it`);
+        }
+
+        if (errorMessage.length) {
+            errorMessage.unshift("Configuration Error: ");
+            this.state.errorMessage = errorMessage.join("\n");
```

Comment:
Evil code! never update state except for the constructor. Else the re-render will not be triggers. 
And don`t change your state in the rendering, that will result in loops.

Always use `setSate`
```
this.setState({
      date: new Date()
});
```

This case, we might not use `errorMessage` as state you can use private memeber